### PR TITLE
Add aci_bd_dhcp_label.py

### PR DESCRIPTION
**Pull request by sig9org**
##### SUMMARY

This module is supposed to manage a DHCP Relay Label with Cisco ACI.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

aci_bd_dhcp_label

##### ANSIBLE VERSION

```
ansible 2.4.0.0
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```

##### ADDITIONAL INFORMATION

N/A.

### CODE REVIEW

**Comment by dagwieers** [comment]
@sig9org Thanks for the contribution. The module is looking fine, however we do like to see examples in the module documentation, as well as integration tests for the functionality. You can look at **aci_tenant** or **aci_filter** for examples on how to write integration tests.

PS Yes, a lot of the modules we add in v2.4 did not have proper examples or integration tests, but that's because we were on a tight schedule to get them included. I plan to fix this situation the coming weeks for the existing modules.

**Review by dagwieers — APPROVED** [review]
The module looks good. Examples and integration tests missing.

**Comment by sig9org** [comment]
@dagwieers Thanks for your reply. I checked **aci_tenant** and **aci_filter**, then I added some examples for this module. Is this enough?

Regards,

**Comment by jmcgill298** [comment]
`dhcp_label` should also be required when the state is absent or present

**Comment by sig9org** [comment]
@jmcgill298 Thank you for your comment. Sorry, it's my mistake. I fixed up it.

### Before

```python
    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True,
        required_if=[
            ['state', 'absent', ['bd', 'tenant']],
            ['state', 'present', ['bd', 'tenant']],
        ],
    )
```

### After

```python
    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True,
        required_if=[
            ['state', 'absent', ['bd', 'tenant', 'dhcp_label']],
            ['state', 'present', ['bd', 'tenant', 'dhcp_label']],
        ],
    )
```

**Comment by dagwieers** [comment]
@sig9org The integration tests are missing as well, you can find examples here: https://github.com/datacenter/aci-ansible/tree/master/test/integration/targets

**Comment by sig9org** [comment]
@dagwieers Thank you! I will use it for future reference.

**Comment by sig9org** [comment]
@dagwieers Can you merge this code? Is integration test necessary? If this code is not enough, please teach me the points.

Regards,

**Comment by dagwieers** [comment]
We need integration tests to ensure that future changes are not impacting existing functionality.
Integration tests live in **test/integration/targets** and they are basically a set of tasks in a playbook to test correct functioning, idempotency and check-mode.

**Review by jmcgill298 — CHANGES_REQUESTED** [review]
please remove changes to aci.py and pull down the latest. We have refactored this to accommodate some special cases in the URLs, and provided a cleaner interface for the modules.

File: lib/ansible/module_utils/aci.py [hunk 1]
```diff
@@ -66,6 +66,7 @@
     aep=dict(aci_class='infraAttEntityP', mo='infra/attentp-', key='name'),
     ap=dict(aci_class='fvAp', mo='ap-', key='name'),
     bd=dict(aci_class='fvBD', mo='BD-', key='name'),
+    bd_dhcp_label=dict(aci_class='dhcpLbl', mo='dhcplbl-', key='name'),
```
Comment (jmcgill298) [inline lib/ansible/module_utils/aci.py:69]: please remove changes to aci.py and pull down the latest. We have refactored this to accommodate some special cases in the URLs, and provided a cleaner interface for the modules.

File: lib/ansible/modules/network/aci/aci_bd_dhcp_label.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    dhcp_label = module.params['dhcp_label']
+    dhcp_option = module.params['dhcp_option']
+    owner = module.params['owner']
+    state = module.params['state']
+
+    module.params['bd_dhcp_label'] = dhcp_label
+
+    aci = ACIModule(module)
+    aci.construct_url(
+        root_class='tenant', subclass_1='bd', subclass_2='bd_dhcp_label',
```
Comment (jmcgill298) [inline lib/ansible/modules/network/aci/aci_bd_dhcp_label.py:139]:
```
    aci.construct_url(
        root_class=dict(
            aci_class='fvTenant',
            aci_rn='tn-{}'.format(tenant),
            filter_target='(fvTenant.name, "{}")'.format(tenant),
            module_object=tenant,
        ),
        subclass_1=dict(
            aci_class='fvBD',
            aci_rn='BD-{}'.format(bd),
            filter_target='(fvBD.name, "{}")'.format(bd),
            module_object=bd,
        ),
        subclass_2=dict(
            aci_class='dhcpLbl',
            aci_rn='dhcplbl-{}'.format(dhcp_label),
            filter_target='(dhcpLbl.name, "{}")'.format(dhcp_label),
            module_object=dhcp_label,
        ),
    )
```

File: lib/ansible/modules/network/aci/aci_bd_dhcp_label.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            ['state', 'present', ['bd', 'tenant', 'dhcp_label']],
+        ],
+    )
+
+    description = module.params['description']
+    dhcp_label = module.params['dhcp_label']
+    dhcp_option = module.params['dhcp_option']
+    owner = module.params['owner']
+    state = module.params['state']
+
```
Comment (jmcgill298) [inline lib/ansible/modules/network/aci/aci_bd_dhcp_label.py:134]:
```
    tenant = module.params['tenant']`
```

File: lib/ansible/modules/network/aci/aci_bd_dhcp_label.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ['state', 'absent', ['bd', 'tenant', 'dhcp_label']],
+            ['state', 'present', ['bd', 'tenant', 'dhcp_label']],
+        ],
+    )
+
```
Comment (jmcgill298) [inline lib/ansible/modules/network/aci/aci_bd_dhcp_label.py:128]:
```
    bd = module.params['bd']
```

**Review by jmcgill298 — CHANGES_REQUESTED** [review]
ACI was refactored to accommodate new modules and provide a cleaner interface. I have provided the necessary updates for the module.

Also, we still need integration tests. I pushed several last week you can use as examples.

**Comment by sig9org** [comment]
@jmcgill298 Thank you for your reply! I will check your integration test examples.

**Comment by gundalow** [comment]
Hi, we are getting close to the cut off (15th Jan 2018) for new modules for Ansible 2.5. Looks like this is close, though just needs a few minor things before it can be merged, such as tests :)

**Comment by gundalow** [comment]
And a rebase

File: lib/ansible/modules/network/aci/aci_bd_dhcp_label.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+DOCUMENTATION = r'''
+---
+module: aci_bd_dhcp_label
+short_description: Manage DHCP Label on Cisco ACI fabrics (dhcp:Lbl)
+description:
+- Manage DHCP Label on Cisco ACI fabrics (dhcp:Lbl)
+- More information from the internal APIC class
+  I(dhcp:Lbl) at U(https://pubhub-prod.s3.amazonaws.com/media/apic-mim-ref/docs/MO-dhcpLbl.html).
+author:
+- sig9 (sig9@sig9.org)
```
Comment (gundalow) [inline lib/ansible/modules/network/aci/aci_bd_dhcp_label.py:22]: Hi, this should be
`- First Last (@GitHub)`

File: lib/ansible/modules/network/aci/aci_bd_dhcp_label.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    argument_spec = aci_argument_spec
+    argument_spec.update(
+        bd=dict(type='str', aliases=['bd_name']),
+        description=dict(type='str', aliases=['descr']),
+        dhcp_label=dict(type='str'),
+        dhcp_option=dict(type='str'),
+        owner=dict(type='str'),
+        state=dict(type='str', default='present', choices=['absent', 'present', 'query']),
+        tenant=dict(type='str', aliases=['tenant_name']),
+        method=dict(type='str', choices=['delete', 'get', 'post'], aliases=['action'], removed_in_version='2.6'),  # Deprecated starting from v2.6
```
Comment (gundalow) [inline lib/ansible/modules/network/aci/aci_bd_dhcp_label.py:117]: No point adding code to deprecate it in the future, please remove.

Comment (gundalow) [reply lib/ansible/modules/network/aci/aci_bd_dhcp_label.py:117]: Doesn't seem to be used anyway...

File: lib/ansible/modules/network/aci/aci_bd_dhcp_label.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+#
+'''
+
+from ansible.module_utils.aci import ACIModule, aci_argument_spec
+from ansible.module_utils.basic import AnsibleModule
+
+
+def main():
+    argument_spec = aci_argument_spec
+    argument_spec.update(
```
Comment (gundalow) [inline lib/ansible/modules/network/aci/aci_bd_dhcp_label.py:109]: Missing `required=True` from a lot of these. This should match `DOCUMENTATION`

File: lib/ansible/modules/network/aci/aci_bd_dhcp_label.py (hunk 6 above)
Comment (dagwieers) [reply lib/ansible/modules/network/aci/aci_bd_dhcp_label.py:117]: The library can make use of it (and reports deprecation warnings), but you are right, we shouldn't be introducing this in new modules.

The older module had some legacy where we didn't want to break playbooks written before we upstreamed those modules.

**Comment by dagwieers** [comment]
I also wonder what the relation is between this **aci_bd_dhcp_label** and the older (non-upstreamed) **aci_bd_dhcp_association** module.

File: lib/ansible/modules/network/aci/aci_bd_dhcp_label.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+RETURN = r'''
+#
+'''
+
+from ansible.module_utils.aci import ACIModule, aci_argument_spec
+from ansible.module_utils.basic import AnsibleModule
+
+
+def main():
+    argument_spec = aci_argument_spec
```
Comment (sivel) [inline lib/ansible/modules/network/aci/aci_bd_dhcp_label.py:108]: `aci_argument_spec` is now a function and needs to be called like: `argument_spec = aci_argument_spec()`

**Comment by gundalow** [comment]
Rebase needed 
https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_5.rst
> Community Module Freeze: 7 February 2018

**Comment by brunocalogero** [comment]
hey @sig9org are you still doing this? I would love to assist you, i've made a few modules already, I can help get this done

File: lib/ansible/modules/network/aci/aci_bd_dhcp_label.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    description:
+    - The name of the Tenant.
+    aliases: [ tenant_name ]
+'''
+
+EXAMPLES = r''' # '''
+
+RETURN = r''' # '''
+
+from ansible.module_utils.aci import ACIModule, aci_argument_spec
```
Comment (mtorelli) [inline lib/ansible/modules/network/aci/aci_bd_dhcp_label.py]: line 54 should be...

from ansible.module_utils.network.aci.aci import ACIModule, aci_argument_spec

this appears to be why the sanity checks are failing

**Comment by dagwieers** [comment]
@rsmeyers What would be the right name ? **aci_bd_dhcp_label** (this PR) or **aci_bd_dhcp_association** (older ACI modules) or something else ?

File: lib/ansible/modules/network/aci/aci_bd_dhcp_label.py (hunk 9 above)
Comment (dagwieers) [reply lib/ansible/modules/network/aci/aci_bd_dhcp_label.py]: Yes, that file has moved in the meantime. That's also why we have that conflict.

**Comment by mtorelli** [comment]
aci_bd_dhcp_label seems appropriate since the module "creates" a DHCP label within a bridge-domain (BD).

**Comment by smnmtzgr** [comment]
@sig9org / @dagwieers is this module ready? If not I can assist. Please tell me whats the actual state here.

**Comment by amitswt** [comment]
> aci_bd_dhcp_label

Is this module available in ansible 2.7 to attach dhcp relay to BD ?

**Comment by dagwieers** [comment]
@rsmeyers @devarshishah3 Please advise. Is this the correct name for the module? Does it offer everything that you expect?

**Comment by rsmeyers** [comment]
name is good, module looks complete

**Comment by dagwieers** [comment]
@rsmeyers OK, I will rework this PR so it is acceptable to merge.

**Comment by gundalow** [comment]
@sig9org Hi, looks like your branch has been deleted from your fork. 
It may be easier for you to create a new branch & PR and copy the two files in.

**Comment by gundalow** [comment]
@dagwieers Hi, is it still on your list to rework this?

**Pull request closed without merging**